docs/usage.rst

```diff
@@ -1,213 +1,233 @@
+
+Global log properties
+---------------------
+
+SeqLog can also add static properties to each log entry that is sent to Seq.
+By default, the following properties are added:
+
+* ``MachineName`` The local machine's fully-qualified host name.
+* ``ProcessId`` The current process Id.
+
+To configure global log properties, call ``set_global_log_properties``, passing the properties as keyword arguments:
+
+.. code-block:: python
+
+    import seqlog
+
+    seqlog.set_global_log_properties(
+        GlobalProperty1="foo",
+        GlobalProperty2="bar"
+        GlobalProperty3=26
+    )
+
+Note that you can also clear the global log properties (so no properties are added) by calling ``clear_global_log_properties``, and reset the global log properties to their defaults by calling ``reset_global_log_properties``.
+
+Note that is you specify a callable as part of global log properties, it will be called
+with no arguments right before logging:
+
+.. code-block:: python
+
+    import seqlog
+
+    def get_trace_id():
+        if tracer.active_span is not None:
+            return hex(tracer.active_span.context.trace_id)
+        else:
+            return None
+
+    seqlog.set_global_log_properties(
+        trace_id=get_trace_id,
+    )
+
+If the callable returns None, it won't be added.
+
+Callback on log submission failure
+----------------------------------
+
+If you wish to set a callable to be invoked each time log submission fails, 
+use the following function:
+
+.. code-block:: python
+
+    from seqlog import set_callback_on_failure
+    
+    def handle_a_failure(e):    # type: (requests.RequestException) -> None
+        print('Failure occurred during log submission: %s' % (e, ))
+        
+   set_callback_on_failure(handle_a_failure)
+
+The callable that you provide will accept a single positional argument, 
+which is the requests exception instance that was the reason for the fail.
+
+.. note:: This callable will be called only for I/O errors, errors stemming
+          from seqlog not being able to convert your records into JSON won't
+          show up here!
+
+Passing in exceptions
+---------------------
+
+There's a couple of ways you can pass in exception information. The only field sent to Seq will be called Exception and it
+may come from a couple of places, in order:
+
+1. A previously rendered exception was cached
+
+2. If FeatureFlag.STACK_INFO is enabled, :code:`record` is set and :code:`record` is not set, it
+   will be taken from :code:`stack_info`
+
+3. If :code:`exc_info` is a tuple then
+    a. If it's first element is None, FeatureFlag.STACK_INFO is enabled and :code:`record.stack_info` is available,
+       :code:`stack_info` will be used to construct the message
+    b. Else formatted :code:`exc_info` will be used
+
+4. If only :code:`exc_info` is given, and it's an Exception, then the stack trace will be attached.
+
+So if you provide both :code:`exc_info` and :code:`stack_info` the code will behave in a weird way.
```

Comment:
Can I suggest “the code will behave in a weird way” -> “behaviour is undefined”?

Reply:
It'd be a lie. It's not so much undefined as hard to put into words.